File: src/portfolio/generate_resume.py
from pathlib import Path
from subprocess import run
from datetime import date
import re

from docx.document import Document
from docx import Document as CreateDocument
from docx.shared import Inches, Pt
from docx.enum.text import WD_ALIGN_PARAGRAPH, WD_TAB_ALIGNMENT

from data import Portfolio, Education, str_date, DocxConfig, Margins
from docx_utils import (
    insertHR,
    set_global_font,
    get_usable_width,
    add_hyperlink,
    add_run,
    ITALIC,
    BOLD,
)


def add_education(doc: Document, education: Education, docx_config: DocxConfig):
    if education.current is False:
        return
    p = doc.add_paragraph()
    text_size = docx_config.text_font_size
    if education.degree is None:
        raise ValueError("No Degree Specified for Education")
    if education.major is None:
        raise ValueError("No Major Specified for Education")
    if education.coursework is None:
        raise ValueError("No Coursework Specified for Education")
    minor = ", " + education.minor + " Minor" if education.minor is not None else ""

    add_run(
        p,
        f"{education.degree} -- {education.major}{minor}\n",
        Pt(text_size),
        BOLD,
    )
    add_run(p, education.name + ", " + education.location + "\n", Pt(text_size))

    if education.expected is not None:
        add_run(
            p,
            f"{str_date(education.start)} - Expected {str_date(education.expected)}\n",
            Pt(text_size),
            ITALIC,
        )
    elif education.end is not None:
        add_run(
            p,
            f"{str_date(education.start)} - {str_date(education.end)}\n",
            Pt(text_size),
            ITALIC,
        )
    add_run(p, f"GPA: {str(education.GPA)}\n", Pt(text_size))
    add_run(p, "Relevant Coursework: ", Pt(text_size), ITALIC)
    add_run(p, ", ".join(education.coursework), Pt(text_size))


def add_info_and_links(doc: Document, por: Portfolio, docx_config: DocxConfig):
    text_size = docx_config.text_font_size
    p = doc.add_paragraph()
    width = get_usable_width(doc)
    p.paragraph_format.tab_stops.add_tab_stop(Inches(width), WD_TAB_ALIGNMENT.RIGHT)
    email = por.info.email
    phone = por.info.phone
    linkedin = por.info.linkedin
    github = por.info.github
    website = por.info.website
    add_run(p, "Website: ", Pt(text_size))
    add_hyperlink(p, website, website, Pt(text_size))
    add_run(p, "\tEmail: ", Pt(text_size))
    add_hyperlink(p, f"{email}\n", f"mailto:{email}", Pt(text_size))
    add_run(p, "LinkedIn: ", Pt(text_size))
    add_hyperlink(p, linkedin, linkedin, Pt(text_size))
    add_run(p, f"\tLocation: {por.info.location}\n", Pt(text_size))

    add_run(p, "GitHub: ", Pt(text_size))
    add_hyperlink(p, github, github, Pt(text_size))
    add_run(p, "\tPhone: ", Pt(text_size))
    add_hyperlink(p, phone, f"tel:{phone}", Pt(text_size))


def add_experiences(doc: Document, por: Portfolio, docx_config: DocxConfig):
    text_size = docx_config.text_font_size
    p = doc.add_paragraph()
    p.add_run("Experience & Projects").font.size = Pt(docx_config.heading_font_size)
    insertHR(p)

    experiences = por.experience
    if experiences is not None:
        for exp in experiences:
            if exp.current is False:
                continue
            p = doc.add_paragraph()
            l1 = exp.company_name
            if exp.location is not None:
                l1 += ", " + exp.location
            l1 += "\n"
            if exp.title_link is not None:
                _, j = add_hyperlink(p, l1, exp.title_link, Pt(text_size))
            else:
                j = p.add_run(l1)
            j.font.size = Pt(text_size)
            j.bold = j.underline = True
            if exp.title is not None:
                add_run(p, exp.title + "\n", Pt(text_size))
            time = str_date(exp.start) + " - "
            if exp.end is None:
                time += "Present"
            else:
                time += str_date(exp.end)
            add_run(p, time, Pt(text_size), ITALIC)
            for line in exp.description:
                doc.add_paragraph(line, "List Bullet")


def add_publications(doc: Document, por: Portfolio, docx_config: DocxConfig):
    text_size = docx_config.text_font_size
    p = doc.add_paragraph()
    add_run(p, "Publications", Pt(docx_config.heading_font_size))
    insertHR(p)

    if por.publications is not None:
        for pub in por.publications:
            p = doc.add_paragraph("", "List Number")
            for i, author in enumerate(pub.authors):
                if i < len(pub.authors) - 1:
                    delimiter = ", "
                else:
                    delimiter = "\n"
                r = add_run(p, f"{author}{delimiter}", Pt(text_size))
                if author == por.info.name:
                    r.bold = True
            pub_str = pub.title
            if "published" not in pub.status.lower():
                pub_str += f"\n{pub.status}"
            add_run(p, pub_str, Pt(text_size))


def add_skills(doc: Document, por: Portfolio, docx_config: DocxConfig):
    text_size = docx_config.text_font_size
    p = doc.add_paragraph()
    add_run(p, "Technical Skills", Pt(docx_config.heading_font_size))
    insertHR(p)

    p = doc.add_paragraph()
    if por.skills is not None:
        first = True
        for name in por.skills.keys():
            if not first:
                p.add_run("\n")
            skill_group = por.skills[name]
            name: str = re.sub("[^0-9a-zA-Z]+", " ", name).capitalize()
            name = f"{name}: "
            add_run(p, name, Pt(text_size), BOLD)
            add_run(p, ", ".join(skill_group), Pt(text_size))
            first = False


def add_awards(doc: Document, por: Portfolio, docx_config: DocxConfig):
    text_size = docx_config.text_font_size
    width = get_usable_width(doc)
    p = doc.add_paragraph()
    add_run(p, "Awards & Certifications", Pt(docx_config.heading_font_size))
    insertHR(p)

    p = doc.add_paragraph()
    p.paragraph_format.tab_stops.add_tab_stop(Inches(width), WD_TAB_ALIGNMENT.RIGHT)
    if por.awards is not None:
        for award in por.awards:
            if award.current is False:
                continue
            add_run(p, f"{award.name}\t", Pt(text_size), BOLD)
            if isinstance(award.date, date):
                date_str = str_date(award.date)
            else:
                date_str = award.date
            add_run(p, f"{date_str}\n", Pt(text_size), ITALIC)


def create_docx(path: str | Path, docx_config: DocxConfig):
    doc = CreateDocument()
    sections = doc.sections
    for section in sections:
        section.top_margin = Inches(docx_config.margins.top)
        section.bottom_margin = Inches(docx_config.margins.bottom)
        section.left_margin = Inches(docx_config.margins.left)
        section.right_margin = Inches(docx_config.margins.right)

    ROOT = Path(__file__).parent.absolute()

    with open(ROOT / "portfolio.json") as f:
        por = Portfolio.model_validate_json(f.read())

    # Add Title
    p = doc.add_paragraph()
    p.alignment = WD_ALIGN_PARAGRAPH.CENTER
    r = p.add_run(por.info.name)
    r.font.size = Pt(docx_config.title_font_size)

    # Info and Links section at the top
    add_info_and_links(doc, por, docx_config)

    # Add education section
    p = doc.add_paragraph()
    p.add_run("Education").font.size = Pt(docx_config.heading_font_size)
    insertHR(p)
    if por.education is not None:
        add_education(doc, por.education[0], docx_config)

    # Experience Section
    add_experiences(doc, por, docx_config)

    # Publications Section
    add_publications(doc, por, docx_config)

    # Skills Section
    add_skills(doc, por, docx_config)

    # Certs Section
    add_awards(doc, por, docx_config)

    set_global_font(doc, "Calibri")
    doc.save(str(path))


def docx_to_pdf(docx_file: str | Path):
    run(
        [
            "libreoffice",
            "--headless",
            "--convert-to",
            "pdf",
            str(docx_file),
            "--outdir",
            str(Path(docx_file).parent),
        ]
    )


if __name__ == "__main__":
    ROOT = Path(__file__).parent.parent.parent.absolute()
    docx_config = DocxConfig(
        margins=Margins(top=0.75, bottom=0.75, left=0.75, right=0.75),
        title_font_size=20,
        heading_font_size=14,
        text_font_size=11,
    )
    create_docx(ROOT / "docs/source/_static/resume.docx", docx_config)
    docx_to_pdf(ROOT / "docs/source/_static/resume.docx")

File: portfolio/src/resume/generate_resume.rs
use docx_rs::*;
use std::process::Command;
use std::{fs::canonicalize, path::Path};

use crate::{
    data::{load_portfolio, EndingDate, Portfolio},
    hyperlink_run, inches_to_twips, PTrait,
};

pub const TITLE_SIZE: usize = 20 * 2;
pub const HEADING_SIZE: usize = 14 * 2;
pub const TEXT_SIZE: usize = 11 * 2;
pub const LINE_SPACING: f32 = 1.15;
pub const LINE_SPACING_TWIPS: i32 = ((TEXT_SIZE / 2) as f32 * LINE_SPACING * 20f32) as i32;

pub fn add_info_and_links(mut doc: Docx, por: &Portfolio) -> Docx {
    let p = Paragraph::new()
        .add_tab_stop()
        .add_run(Run::new().add_text("Website: "))
        .add_hyperlink(
            Hyperlink::new(por.info.website.clone(), HyperlinkType::External)
                .add_run(hyperlink_run(por.info.website.clone())),
        )
        .add_run(Run::new().add_text("\tEmail: "))
        .add_hyperlink(
            Hyperlink::new(
                &format!("mailto:{}", por.info.email.clone()),
                HyperlinkType::External,
            )
            .add_run(hyperlink_run(por.info.email.clone()).add_break(BreakType::TextWrapping)),
        )
        .add_run(Run::new().add_text("LinkedIn: "))
        .add_hyperlink(
            Hyperlink::new(por.info.linkedin.clone(), HyperlinkType::External)
                .add_run(hyperlink_run(por.info.linkedin.clone())),
        )
        .add_run(Run::new().add_text("\tLocation: "))
        .add_run(
            Run::new()
                .add_text(por.info.location.clone())
                .add_break(BreakType::TextWrapping),
        )
        .add_run(Run::new().add_text("Github: "))
        .add_hyperlink(
            Hyperlink::new(por.info.github.clone(), HyperlinkType::External)
                .add_run(hyperlink_run(por.info.github.clone())),
        )
        .add_run(Run::new().add_text("\tPhone: "))
        .add_hyperlink(
            Hyperlink::new(
                &format!("tel:{}", por.info.phone.clone()),
                HyperlinkType::External,
            )
            .add_run(hyperlink_run(por.info.phone.clone())),
        );
    // doc = &mut doc.add_paragraph(p);
    doc = doc.add_paragraph(p);
    doc
}

fn add_education(mut doc: Docx, por: &Portfolio) -> Docx {
    doc = doc.add_paragraph(
        Paragraph::new()
            .add_run(Run::new().add_text("Education").size(HEADING_SIZE))
            .insert_hr(),
    );
    for edu in &por.education {
        if !edu.current {
            continue;
        }
        let mut body_str = String::new();
        if let Some(degree) = &edu.degree {
            body_str.push_str(degree);
        }
        if edu.major.is_some() && edu.degree.is_some() {
            body_str.push_str(" -- ");
        }
        if let Some(major) = &edu.major {
            body_str.push_str(major);
        }
        let mut p = Paragraph::new()
            .add_tab(Tab {
                val: Some(TabValueType::Right),
                leader: None,
                pos: Some(inches_to_twips(7) as usize),
            })
            .add_run(Run::new().bold().add_text(body_str))
            .add_run(
                Run::new()
                    .add_text({
                        let mut duration_str = format!("\t{}", edu.duration.start.str_date());
                        match &edu.duration.end {
                            EndingDate::End(end) => {
                                duration_str.push_str(&format!(" - {}", end.str_date()));
                            }
                            EndingDate::Current {
                                expected: Some(date),
                            } => {
                                duration_str.push_str(&format!(" - Expected {}", date.str_date()));
                            }
                            EndingDate::Current { expected: None } => {
                                duration_str.push_str(" - Present");
                            }
                        }
                        duration_str
                    })
                    .italic()
                    .add_break(BreakType::TextWrapping),
            )
            .add_run(
                Run::new()
                    .add_text(&format!("{}, {}", edu.name, edu.location))
                    .add_break(BreakType::TextWrapping),
            );
        if let Some(minor) = &edu.minor {
            p = p.add_run(
                Run::new()
                    .add_text(&format!("Minor: {}", minor))
                    .add_break(BreakType::TextWrapping),
            );
        }
        p = p
            .add_run(
                Run::new()
                    .add_text(&format!("GPA: {:.1}", edu.gpa))
                    .add_break(BreakType::TextWrapping),
            )
            .add_run(Run::new().add_text("Relevant Coursework: ").italic());
        if let Some(coursework) = &edu.coursework {
            p = p.add_run(Run::new().add_text(coursework.join(", ")));
        }
        doc = doc.add_paragraph(p);
    }
    doc
}

fn add_experiences(mut doc: Docx, por: &Portfolio) -> Docx {
    doc = doc.add_paragraph(
        Paragraph::new()
            .add_run(
                Run::new()
                    .add_text("Experience & Projects")
                    .size(HEADING_SIZE),
            )
            .insert_hr(),
    );
    for exp in &por.experience {
        if !exp.current {
            continue;
        }
        let mut p = Paragraph::new().add_tab_stop();
        let mut first_line = exp.company_name.clone();

        if let Some(location) = &exp.location {
            first_line.push_str(&format!(", {location}"));
        }
        if let Some(title_link) = &exp.title_link {
            let run = hyperlink_run(first_line).bold();
            p = p.add_hyperlink(Hyperlink::new(title_link, HyperlinkType::External).add_run(run))
        } else {
            let run = Run::new().add_text(first_line).bold().underline("single");
            p = p.add_run(run);
        }

        let mut time_str = format!("\t{} - ", exp.duration.start.str_date());
        time_str.push_str({
            match &exp.duration.end {
                EndingDate::End(end) => end.str_date(),
                EndingDate::Current {
                    expected: Some(exp),
                } => format!("Expected {}", exp.str_date()),
                EndingDate::Current { expected: None } => "Present".into(),
            }
            .as_str()
        });
        let mut r = Run::new().add_text(time_str).italic();
        if exp.title.is_some() {
            r = r.add_break(BreakType::TextWrapping);
        }
        p = p.add_run(r);

        if let Some(title) = &exp.title {
            p = p.add_run(Run::new().add_text(title));
        }

        doc = doc.add_paragraph(p);
        for line in &exp.description {
            doc = doc.add_paragraph(
                Paragraph::new()
                    .add_run(Run::new().add_text(line))
                    .numbering(NumberingId::new(8), IndentLevel::new(0)),
            );
        }
    }
    doc
}

fn add_publications(mut doc: Docx, por: &Portfolio) -> Docx {
    doc = doc.add_paragraph(
        Paragraph::new()
            .add_run(Run::new().add_text("Publications").size(HEADING_SIZE))
            .insert_hr(),
    );
    for publication in &por.publications {
        let mut p = Paragraph::new().numbering(NumberingId::new(1), IndentLevel::new(0));
        for (i, author) in publication.authors.iter().enumerate() {
            let mut r = if i < publication.authors.len() - 1 {
                Run::new().add_text(format!("{author}, "))
            } else {
                Run::new()
                    .add_text(author)
                    .add_break(BreakType::TextWrapping)
            };
            if *author == por.info.name {
                r = r.bold();
            }
            p = p.add_run(r);
        }
        let mut pub_str_r = Run::new().add_text(publication.title.clone());
        if !publication.status.to_lowercase().contains("published") {
            pub_str_r = pub_str_r.add_break(BreakType::TextWrapping);
            p = p.add_run(pub_str_r);
            p = p.add_run(Run::new().add_text(publication.status.clone()));
        } else {
            p = p.add_run(pub_str_r);
        }
        doc = doc.add_paragraph(p);
    }
    doc
}

pub fn add_skills(mut doc: Docx, por: &Portfolio) -> Docx {
    doc = doc.add_paragraph(
        Paragraph::new()
            .add_run(Run::new().add_text("Technical Skills").size(HEADING_SIZE))
            .insert_hr(),
    );
    for skills in &por.skills {
        let mut p = Paragraph::new().numbering(NumberingId::new(8), IndentLevel::new(0));
        let group = skills.group_name.clone();
        p = p.add_run(Run::new().add_text(&format!("{group}: ")).bold());
        p = p.add_run(Run::new().add_text(skills.skills.join(", ")));
        doc = doc.add_paragraph(p);
    }
    doc
}

pub fn add_awards(mut doc: Docx, por: &Portfolio) -> Docx {
    doc = doc.add_paragraph(
        Paragraph::new()
            .add_run(
                Run::new()
                    .add_text("Awards & Certifications")
                    .size(HEADING_SIZE),
            )
            .insert_hr(),
    );
    for awd in &por.awards {
        let mut p = Paragraph::new().add_tab_stop();
        if let Some(link) = &awd.link {
            p = p
                .add_hyperlink(
                    Hyperlink::new(link, HyperlinkType::External).add_run(
                        Run::new()
                            .add_text(awd.name.clone())
                            .color("0563C1")
                            .underline("single"),
                    ),
                )
                .add_run(Run::new().add_text(&format!("\t{}", awd.date.str_date())));
        } else {
            p = p.add_run(Run::new().add_text(&format!("{}\t{}", awd.name, awd.date.str_date())));
        }
        doc = doc.add_paragraph(p);
    }
    doc
}

pub fn generate_resumes(path: &Path) {
    let por = load_portfolio();
    let _ = std::fs::create_dir_all(path.parent().unwrap());
    let file = std::fs::File::create(path).unwrap();
    let p = Paragraph::new()
        .align(AlignmentType::Center)
        .add_run(Run::new().add_text(por.info.name.clone()).size(TITLE_SIZE));
    let mut doc = Docx::new()
        .page_size(inches_to_twips(8.5) as u32, inches_to_twips(11) as u32)
        .page_margin(PageMargin {
            top: inches_to_twips(0.75),
            left: inches_to_twips(0.75),
            bottom: inches_to_twips(0.75),
            right: inches_to_twips(0.75),
            header: 0,
            footer: 0,
            gutter: 0,
        })
        .default_line_spacing(LineSpacing::new().line(LINE_SPACING_TWIPS))
        .default_size(TEXT_SIZE)
        .default_fonts(RunFonts::new().ascii("Carlito"))
        .add_abstract_numbering(
            AbstractNumbering::new(8).add_level(
                Level::new(
                    0,
                    Start::new(1),
                    NumberFormat::new("bullet"),
                    LevelText::new("•"),
                    LevelJc::new("left"),
                )
                .indent(
                    Some(360),
                    Some(SpecialIndentType::Hanging(360)),
                    None,
                    None,
                ),
            ),
        )
        .add_numbering(Numbering::new(8, 8))
        .add_paragraph(p);

    doc = add_info_and_links(doc, &por);

    doc = add_education(doc, &por);
    doc = add_experiences(doc, &por);
    doc = add_publications(doc, &por);
    doc = add_skills(doc, &por);
    doc = add_awards(doc, &por);
    doc.build().pack(file).expect("Error creating docx");
    docx_to_pdf(path);
}

pub fn docx_to_pdf(path: &Path) {
    let path = canonicalize(path).unwrap();
    let _ = Command::new("libreoffice")
        .args([
            "--headless",
            "--convert-to",
            "pdf",
            path.to_str().unwrap(),
            "--outdir",
            path.parent().unwrap().to_str().unwrap(),
        ])
        .output()
        .expect("Failed to convert docx to pdf");
}
